src/chapters/loops/while-loops.rst

```diff
@@ -0,0 +1,199 @@
+#. If the condition is ``false``, exit the ``while`` loop and continue execution at the next statement after the loop body.
+#. If the condition is ``true``, execute the loop body and then go back to step 1.
+
+**for** Loops Rewritten as **while** Loops
+------------------------------------------
+
+We can use the ``while`` loop to create any type of iteration we wish, including anything that we have previously done with a ``for`` loop. For example, consider our initial ``for`` loop example.
+
+.. sourcecode:: js
+
+   for (let i = 0; i < 51; i++) {
+      console.log(i);
+   }
+
+This can be rewritten as a while loop:
+
+.. sourcecode:: js
+
+   let i = 0;
+
+   while (i < 51) {
+      console.log(i);
+      i++;
+   }
+
+Instead of relying on the the initial and update expressions, as we do in a ``for`` loop, we must manage the state of our loop manually. To do this, *before* entering the ``while`` loop, we will create the variable ``i`` and initialize it to 0, the first number we want to print. This variable plays the same role as the loop variable in a ``for`` loop. Every iteration will print ``i`` and then increment ``i`` to the next value, until all it reaches the value 51. The loop continues to iterate until the condition ``i < 51`` evaluates to ``false``.
+
+You can almost read the ``while`` statement as if it were in a natural language: *while ``i`` is less than or equal to ``51``, continue executing the body of the loop*.
+
+Beyond **for** Loops
+--------------------
+
+We stated earlier that ``while`` loops are more flexible than ``for`` loops. Now we will look at an example that illustrates this.
+
+This program is an example of **input validation**. It prompts the user to enter a positive number, converting the input string to the number data type. If the number is not positive, then the user is prompted again within the body of the loop. As long as the user continues to input non-positive numbers, the loop will continue to iterate.
+
+.. sourcecode:: js
+
+   const input = require('readline-sync');
+
+   let num = input.question('Please enter a positive number:');
+   num = Number(num);
+
+   while (num <= 0) {
+      num = input.question('Invalid input. Please enter a positive number:');
+      num = Number(num);
+   }
+
+.. index::
+   pair: input; validation
+
+This example illustrates the additional flexibility provided by ``while`` loops. While we use ``for`` loops to iterate over fixed collections (a string, an array, a collection of integers), the ``while`` loop can be used to iterate in more general circumstances. For the input validation example, at runtime it cannot be determined how many times the loop will repeat.
+
+.. warning:: It is *technically* possible to write a ``while`` loop as a ``for`` loop, since ``for`` loops use a condition, and the initial and update expressions can be left empty. Doing so, however, results in confusing code. This should not be done under any circumstances!
```

Comment:
Remove the warning box!  It only serves to confuse the point.  

If you must include a warning, do NOT provide sample code and then tell the students, “Ignore this.”  Instead, focus on the general vs. specific nature of ``while`` and ``for`` loops.  All squares are rectangles, but not all rectangles are squares.  Similarly, a ``while`` loop can perform the same actions as any ``for`` loop, but a ``for`` loop cannot easily perform the same actions as any ``while`` loop.  Discourage trying ``while`` ---> ``for`` translations.

How about, “Given its greater flexibility, a ``while`` loop can be written in place of any ``for`` loop.  However, the reverse is NOT true.  Trying to replace a ``while`` loop with a ``for`` loop requires more work and a confusing code structure, and your code will probably require extensive debugging.  If you find yourself thinking, ‘How can I replace this ``while`` with a ``for``,’ STOP!  Just use the ``while`` loop.”

Reply:
This feels like a topic to work out at a larger scale. This type of thing comes up a lot with JS. We want to steer students in a certain direction without hiding things from them (both because they'll see this stuff online anyway, and because a pro dev would be expected to know many of these things). What approach can we use to hit both of these objectives?

Reply:
Steer without hiding?  Tough call.  Perhaps something like, "Best practice strongly encourages us to avoid _____, but you will probably see such situations as you examine code samples online.  To explore ____ and see how awkward it is, click **here** for examples."  

Then we put the examples (carefully selected to emphasize the point we want to make) in an appendix.  Maybe title it "NOT best-practices".  If our goal is to teach the students how to spot and avoid sloppy code, then the examples should show the nastiness as well as a cleaner option.  Perhaps end the examples with a repl.it that asks the students to convert sloppy code into something more elegant that works better and is more readable.

I have other thoughts that would be easier to discuss rather than type.  Let's chat.